### PR TITLE
ISSUE-703 LabelRepository memory implementation

### PR DESCRIPTION
**Pull request by quantranhong1999**
resolve https://github.com/linagora/tmail-backend/issues/703

### CODE REVIEW

**Review by Arsnael — COMMENTED** [review]
Read it

**Comment by Arsnael** [comment]
```
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile (scala-compile-first) on project jmap-extensions: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile failed: Compilation failed: InterfaceCompileFailed -> [Help 1]
```

**Review by Arsnael — APPROVED** [review]

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/label/LabelRepositoryContract.scala [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+  }
+
+  @Test
+  def deleteShouldBeIdempotent(): Unit = {
+    val randomLabelId = LabelId.generate()
+
+    assertThatCode(() => SMono.fromPublisher(testee.deleteLabel(ALICE, randomLabelId)).block())
+      .doesNotThrowAnyException()
+  }
+}
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/label/LabelRepositoryContract.scala:236]: Should we?
- updateShouldNotThrowWhenIndempotent
- updateShouldThrowWhenBotColorAndDisplayNameIsEmpty

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/Label.scala [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  def toKeyword: Keyword =
+    Keyword.of(id.value).get
+}
+
+object KeywordUtil {
+  def generate(): Keyword =
+    Keyword.of(UUID.randomUUID().toString).get
+}
+
+case class DisplayName(value: String)
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/Label.scala:27]: NonEmpty type

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/Label.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+case class Color(value: String)
```
[the remark this answers is not in the document]
Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/Label.scala:29]: > Should we create a Color type with ColorConstraint? 

Yes, we should. But IMO that validation could be added upon Color object creation (e.g Label/set create). Here at the repository layer (assuming we already have a valid Color object), we just care about persisting it into DB. 

Is that ok with you?

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/label/LabelRepositoryContract.scala (hunk 1 above)
Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/label/LabelRepositoryContract.scala:236]: > updateShouldThrowWhenBothColorAndDisplayNameIsEmpty

Why does it should throw? Maybe just not overriding anything is good enough? e.g: https://github.com/linagora/tmail-backend/pull/724/commits/856c110ab3e0fe8de33106fa8a7d848a9403b0e9

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/Label.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (Arsnael) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/Label.scala:29]: > My copilot suggests regex for validating input is color:

Your copilot? :)

[the remark this answers is not in the document]
Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/Label.scala:29]: > Your copilot? :)

He uses Github copilot :p

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/label/LabelRepositoryContract.scala (hunk 1 above)
Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/label/LabelRepositoryContract.scala:266]: that's ok

**Review by vttranlina — APPROVED** [review]

**Pull request merged**